### PR TITLE
Issue 5527: Restoring empty segments.

### LINKED ISSUE #5527: In DR tests, CreateReaderGroup() API fails post restart of components
**Problem description**
On running test `RestoreBackupDataRecoveryTest.testDurableDataLogFailRecoveryWatermarking()` this is what I notice:
1. The createScope() API in the test completes successfully and I see creation logs for all 3 metadata tables this API is expected to create.
2. If we issue `createReaderGroup()` API, prior to restarting components, as is done in the test post restart, it completes successfully without any errors. 
3. Post restart the "createReaderGroup()" API fails because the required table segment is not found.
4.  The fix in PR 5517, just recreates the table if it is not found and thus keeps the test from failing.

But the issue here is that some segments are missing post restart.

**Problem location**
Segment Store

**Suggestions for an improvement**

### PR DESCRIPTION
**Pull request by ManishKumarKeshri**
**Change log description**  
During update attributes, if it's found that one segment exists in the old metadata segment, but doesn't exist in the new metadata segment (one created after recovery), then register the segment using the details in the old container metadata segment and update its attrubutes.

**Purpose of the change**  
Fixes #5527 

**What the code does**  
Objective is to register the segments which don't exist in the LTS, but are present in container metadata segment. 

**How to verify it**  
Build shall pass.

### CODE REVIEW

**Review by sachin-j-joshi — APPROVED** [review]
LGTM

**Comment by ManishKumarKeshri** [comment]
One if the tests `testDurableDataLogFailRecoveryWatermarking ` is flaky. Looking into it.

**Review by sachin-j-joshi — CHANGES_REQUESTED** [review]
ok

**Comment by ManishKumarKeshri** [comment]
> ok

flakiness will be resolved in this PR: https://github.com/pravega/pravega/pull/5816

**Review by andreipaduroiu — CHANGES_REQUESTED** [review]

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerRecoveryUtils.java [hunk 1]
```diff
@@ -373,6 +374,35 @@ public static void updateCoreAttributes(Map<Integer, String> backUpMetadataSegme
         }
     }
 
+    /**
+     * * This method takes a {@link DebugStreamSegmentContainer} instance and a {@link SegmentProperties} object as arguments
+     *      * and takes one of the following actions:
```
[the remark this answers is not in the document]
Comment (ManishKumarKeshri) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerRecoveryUtils.java:379]: That was supposed to be deleted. Fixed it.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerRecoveryUtils.java [hunk 2]
```diff
@@ -373,6 +374,35 @@ public static void updateCoreAttributes(Map<Integer, String> backUpMetadataSegme
...
+     *      * 1. If the segment is present in the {@link MetadataStore} of the container and its length or sealed status or both
+     *      * doesn't match with the corresponding details from the given {@link SegmentProperties}, then it is deleted from there
+     *      * and registered using the details from the given {@link SegmentProperties} instance.
+     *      * 2. If the segment is absent in the {@link MetadataStore}, then it is registered using the details from the given
+     *      * {@link SegmentProperties}.
+     *
+     * Registers a segment in the {@link MetadataStore} of the given {@link DebugStreamSegmentContainer} instance using the
```
[the remark this answers is not in the document]
Comment (ManishKumarKeshri) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerRecoveryUtils.java:386]: Using just this one.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerRecoveryUtils.java [hunk 3]
```diff
@@ -373,6 +374,35 @@ public static void updateCoreAttributes(Map<Integer, String> backUpMetadataSegme
...
+     *                      If the operation failed, the future will be failed with the causing exception.
+     */
+    protected static CompletableFuture<Void> registerSegmentAndUpdateAttributes(DebugStreamSegmentContainer container,
+                                                                            SegmentProperties segment, Duration timeout) {
+        // Get the attributes for the segment
+        List<AttributeUpdate> attributeUpdates = segment.getAttributes().entrySet().stream()
+                .map(e -> new AttributeUpdate(e.getKey(), AttributeUpdateType.Replace, e.getValue()))
+                .collect(Collectors.toList());
+        return container.registerSegment(segment.getName(), segment.getLength(), segment.isSealed())
+                .thenAccept(x -> container.updateAttributes(segment.getName(), attributeUpdates, timeout));
```
[the remark this answers is not in the document]
Comment (ManishKumarKeshri) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerRecoveryUtils.java:403]: Ok! Done.

**Review by andreipaduroiu — APPROVED** [review]

**Comment by ManishKumarKeshri** [comment]
!Please do not merge. The added test `testRegisterSegmentAndUpdateAttributes` appears to be flaky. The problem is:
Though I create a segment with deleted = false [here](https://github.com/pravega/pravega/blob/1280b96fb0962c98f61b5d75b21101d5a3a8af58/segmentstore/server/src/test/java/io/pravega/segmentstore/server/containers/DebugStreamSegmentContainerTests.java#L521), but finally when it tries to get the segment info [here](https://github.com/pravega/pravega/blob/1280b96fb0962c98f61b5d75b21101d5a3a8af58/segmentstore/server/src/test/java/io/pravega/segmentstore/server/containers/DebugStreamSegmentContainerTests.java#L508), sm.isDeleted is returned as true.
```
    private CompletableFuture<SegmentProperties> getSegmentSnapshot(long segmentId, String segmentName) {
        SegmentMetadata sm = this.connector.containerMetadata.getStreamSegmentMetadata(segmentId);
        if (sm == null || sm.isDeleted() || sm.isMerged()) {
            return Futures.failedFuture(new StreamSegmentNotExistsException(segmentName));
        } else {
            return CompletableFuture.completedFuture(sm.getSnapshot());
        }
    } 
```
while creation
```
Name = segment--920768406, StartOffset = 0, Length = 3076, Storage Length = 0, Sealed = false, Deleted = false, Sealed in Storage = false, Deleted in Storage = false
```
while retrieval
```
Id = 3, Start = 0, Length = 3076, StorageLength = 3076, Sealed(M/S) = false/false, Deleted = true, Name = segment--920768406`
```

**Review by RaulGracia — CHANGES_REQUESTED** [review]
Blocking for safety.

**Comment by RaulGracia** [comment]
After a discussion with @andreipaduroiu, it looks like we may not need this one in the short term. Closing it.

**Pull request closed without merging**